src/NumericalAlgorithms/Interpolation/LinearLeastSquares.hpp

```diff
@@ -0,0 +1,57 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <gsl/gsl_multifit.h>
+#include <memory>
+#include <vector>
+
+#include "Utilities/GenerateInstantiations.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace intrp {
+/*!
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief A linear least squares solver class
+ *
+ * A wrapper class for the gsl linear least squares solver which determines
+ * the coefficients of best fit for a polynomial of order `Order` given a set
+ * of data points `x_values` and `y_values` representing some function y(x).
+ *
+ * The coeffcients are stored in the member variable `c_values`.
+ * The details of the linear least squares solver can be seen here:
+ * [GSL documentation](https://www.gnu.org/software/gsl/doc/html/lls.html#).
+ *
```

Comment:
I think it would be good to document why this is implemented as a class rather than a free function. Naively I would've expected it to be a free function. However, I think there's value in using a class to store the memory allocations. 

Why are you using the current implementation instead of something like (I'm not saying one or the other is right, just want to discuss so we can converge on what we think is the best solution!):

- Don't take the `x_values` and `y_values` in the constructor (you're doing an unnecessary copy here!), instead take the number of observations/grid points/whatever you want to call them. Store the size as a member variable.
- Have member variables for all the GSL allocations that you need.
- Have `c_values` (maybe name this `fit_coffecients`?) take `x_values` and `y_values` with an internal `ASSERT` on the size, then do the actual fit.
- Have `operator()` (or maybe an `interpolate` function instead? I don't know if the call operator makes sense intuitively to me for interpolation here) take the `fit_coefficients` and `x_to_interpolate_to` and return the value


Something I don't understand about the current implementation is why you need to store the `x_values` and `y_values` as members. Don't you only need them in the constructor? If that's the case, that would help with excessive copying.

I think the choice of implementation boils down to whether we expect to have code where we have a lot of data of the same length that we need to fit. For example, if you have Ylm coeffs and you do a time interpolation on each coefficient, it'll be better to use my suggested implementation because you only need to allocate the workspace once and then can do repeated fits (assuming GSL doesn't do something super weird, so please test that before deleting your code!). That way we only need to allocate the workspaces, etc. once.